### PR TITLE
Add diagnostics for E0172, fix inline message for E0139

### PR DESCRIPTION
**Pull request by AlisdairO**
As title :-)

Part of #24407.

### CODE REVIEW

**Comment by Manishearth** [comment]
So sorry, I already did this in https://github.com/Manishearth/rust/commit/850e9be9538f4641f7344a382241d5d47afa436e . 

I'll update the spreadsheet to give myself a global lock over librustc

**Comment by Manishearth** [comment]
If you think that one can be improved or this one is better, let me know and we can merge them :)

**Comment by AlisdairO** [comment]
Ah that's a shame :(.  Oh well - is E0172 still okay?  I could pull that out into a separate PR.

On 172, I think yours is a better explanation of more common scenarios to hit this problem.  That said, I think the short error message may be a little off?  It talks about type parameters in its interior, but for the following example, I wouldn't really call the parameter 'interior':

```
fn foo<T: std::fmt::Display()>(bar: T) {
    unsafe {
        let baz: std::raw::TraitObject = std::mem::transmute(bar);
    }
}
```

**Comment by Manishearth** [comment]
Sure, in fact just amending the commit in this PR would work too.

Hmm. "Interior" might be alluding to situations where there are no explicit type params involved.

E.g.

```
fn foo<T>(x: *const T) -> Box<T>{
type ConcreteRaw = *const T;
type Concrete = Box<T>;
transmute::<ConcreteRaw, Concrete>(x);
}
```

But this isn't an important distinction, really. "contaisn unsubstituted type parameters" or some such might be better. (Feel free to make the change to the inline error in this same PR)

**Comment by AlisdairO** [comment]
OK, I think that should do it!

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -1476,6 +1476,33 @@ return, for example with a `loop` that never breaks or a call to another
...
+E0172: r##"
+This error means that an attempt was made to specify the type of a variable with
+a combination of a concrete type and a trait. Consider the following example:
+
+```
+fn foo(bar: i32+std::fmt::Display) {}
+```
+
+The code is trying to specify that we want to receive a signed 32-bit integer
+which also implements Display. This doesn't make sense: when we pass `i32`, a
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: nit: backticks around Display

**Comment by Manishearth** [comment]
LGTM modulo nit

**Comment by AlisdairO** [comment]
Added the nit-fix, and squashed.

**Comment by GuillaumeGomez** [comment]
@AlisdairO: If you wanna see what error code remains, take a look at the spreadsheet available here : #24407.

**Comment by AlisdairO** [comment]
@GuillaumeGomez Thanks a lot - I've been going through and marking them off in the spreadsheet, but I think we just hit an unfortunate coincidence :-)

**Comment by GuillaumeGomez** [comment]
On which one ?

**Comment by AlisdairO** [comment]
On E0139.

**Comment by GuillaumeGomez** [comment]
What's about E0139 ? It was marked as merged ?

**Comment by AlisdairO** [comment]
I don't think so - at least, as I see the spreadsheet I'm still marked as the assignee.  I see from the issue that I'm supposed to comment in there too, which I haven't been doing, so I'll do that in future!

**Pull request merged**